### DISCUSSION TITLE
How to generate keystore? in recent versions?

### DISCUSSION THREAD

**Discussion by krnbr** [panva/jose#335] [Q&A] [answered]
How to generate keystore? in recent versions?

**Comment by panva** [comment, marked as answer]
There is no keystore structure in v3/v4

**Reply by krnbr** [reply]
Please support in some newer versions

**Reply by panva** [reply]
There's no reason to.

**Reply by krnbr** [reply]
so what's the way for doing that using panva/jose?? without keystore how to do that?

**Reply by panva** [reply]
do --> what <--?

**Reply by krnbr** [reply]
create array of keys(keypair) 

use any of those keypair to sign JWT(s)

and export the public keys as jwks for validating the signature

**Reply by panva** [reply]
There's a function for generating key pairs.
There's a function for exporting the keys as JWK.
JWKS structure is a simple object with a keys array property.
There's a function for calculating the unique JWK Thumbprint to use as `kid`.

**Reply by krnbr** [reply]
how to get kids from the keys?? do we need to call the `jose.calculateJwkThumbprint`

**Reply by panva** [reply]
`kid` can be an arbitrary string, but to use the JWK thumbprint as a `kid` use the aforementioned `calculateJwkThumbprint`.

**Reply by krnbr** [reply]
```
let g1 = await jose.generateKeyPair('RS256');
const privateJwk = await jose.exportJWK(g1.privateKey)
const publicKey = await jose.exportJWK(g1.publicKey)
privateJwk.use = 'sig'
privateJwk.kid = await jose.calculateJwkThumbprint(privateJwk);

```
Like above?

**Reply by krnbr** [reply]
same way kid attribute for the public key also? correct?

Why can't we have a common util function to supply us pre filled kid, use attributes? SUGGESTION + QUESTION!!

**Reply by panva** [reply]
> Why can't we have a common util function to supply us pre filled kid?

Because it's as simple as assigning the property to an object and there are use cases where automatically adding a `kid` just results in an unnecessary calculation.

> same way kid attribute for the public key also? correct?

JWK Thumbprint is only calculated from the public key properties for obvious reasons.

**Reply by krnbr** [reply]
Otherwise Thanks a lot for proactively replying back
